### ISSUE TITLE
Level 3 Server Rest

### ISSUE THREAD

**Issue by pikawrex** [nus-cs2030/2122-s1#272]
When the server is resting, would there be any customers in the queue? For example, given that all the other servers are taken and the last server is resting but has no one in the queue, would the customer leave or wait there?

**Comment by JordanChua** [comment]
As how a customer would choose the server to either get served by or wait at, the customer first scans from the 1st to the last server and check if any server is available to server them. If not, it scans starting from the 1st server to see if there is space in the queue for him to wait, regardless of whether or not the server is resting. So to answer your question the customer will indeed wait at the last server if all other servers are currently serving a customer/resting and they have no space in their queue for the customer to wait :)

**Comment by pikawrex** [comment]
Hi Jordan, thanks for the reply! My main confusion is because i'm not sure what happened with 9th customer in the 2nd test case for level 3.

According to the input, there isn't supposed to be any resting time after server 1 serves the 9th customer, but there is a delay between the ending serve event for the 9th customer, and the serve event for the 10th customer.

Any help in understanding this test case would be great, thanks!

**Comment by pikawrex** [comment]
Thanks for the clear and concise answer Jordan!

**Comment by Hongyi6328** [comment]
rest just means the server is not currently serving any customer, but it can still accept new customers into the queue given that it has vacancies.
I think most people, including I, were stuck thinking how to delay serving when a server needs to rest.
Here are some approaches, any will work:
1. design a RestEvent and when this event is triggered, update all ServeEvent in the event queue that are associated to the server resting.
2. Let your ServeEvent time to be lazily evaluated, that is, delay the evaluation of the time to serve a customer until the customer is the first in the customer queue.
3. Let the resting server "pretend" that it needs to serve an additional customer when it has done serving the previous customer.

**Comment by albertsutz** [comment]
Hi, just to make sure you completely know the whole picture, remember that:

Let's say there's only 1 server with queue of max size 2

the condition is like this:
```
currentCustomerId = 1
Queues: <2, 3>
```

let's say the customer is finished serving (DONE) and there's a rest time of 10 seconds

```
5.000 1 done serving by server 1
```
then your server condition must be
```
currentCustomerId = nobody
Queues: <2, 3>
```
and when a customer come at time = 10.000, he will leave as the queues is all full

^^
I was stuck at this level because of this thing :)

**Comment by Lokki-ki-ki** [comment]
Hiii to share my ideas and hope can help:
I think when the server is resting, the customer still can arrive and find the suitable server to wait. So for me I created five Events: ArriveEvent, WaitEvent, RestEvent, ServeEvent, DoneEvent, and LeaveEvent. The priority follows this sequence. The DoneEvent calls the RestEvent and set the Server status to be "rest". When the RestEvent is poll(), the status of server change again according to what next event should be (e.g if has next in queue then call ServeEvent and change to "work" status.)

**Comment by wjingen** [comment]
> rest just means the server is not currently serving any customer, but it can still accept new customers into the queue given that it has vacancies. I think most people, including I, were stuck thinking how to delay serving when a server needs to rest. Here are some approaches, any will work:
> 
> 1. design a RestEvent and when this event is triggered, update all ServeEvent in the event queue that are associated to the server resting.
> 2. Let your ServeEvent time to be lazily evaluated, that is, delay the evaluation of the time to serve a customer until the customer is the first in the customer queue.
> 3. Let the resting server "pretend" that it needs to serve an additional customer when it has done serving the previous customer.

Hi @Hongyi6328 ,

With regards to your first proposed solution, will this cause a lot of issues when implementing Level 5? I'm afraid that I'll need to redo the implementation if I do this way. Is there any ideal solution you proposed? Currently stuck on level 3.

Also, just to clarify, this means that I would need something like this right:

`if (event.getType() == "rest" `
 `   for (Event e: queue) { `
`if e.getType() == "serves" && event.getServer() == e.getServer()`
`return new Event(---,"serves"----,updatedServeTime)`

Basically, I create new ServeEvents for every ServeEvent for the server resting, right?

Thanks in advance and sorry about the code formatting!

**Comment by PokezardVGC** [comment]
Hi peeps, just to check if I'm getting the right understanding. For the test cases of rest times, can I assume that when the 1st person arrives (as in time order, not customer 1, etc), he will take the rest time of the 1st entry? Following this test case, is it right that customer 2(since came in 2nd via arrive event) will have his/her server to rest ltr?
![image](https://user-images.githubusercontent.com/71965541/142432636-1a89124f-c04c-4568-a031-ab35fbc9b8ec.png)

**Comment by teominghuang** [comment]
@PokezardVGC Hi, that might not be the case. The resting time is done in chronological order based on whichever customer is done first and is relegated not based on the customer id or arrival sequence. The resting time is dished out based on the chronological order of which servers finish their customer first and are going to enjoy a rest next. I'll give an example to further illustrate my point. 

Assuming that customer 1 has an arrival time of 0.00 and a service time of 3s, and customer 2 has an arrival time of 1.00 and a service time of 0.50s. Customer 1 will get relegated to server 1 and customer 2 to server 2. Assuming that the chronological sequence of the rest time is [0, 3.138742] accordingly. Since server 2 will complete customer 2 (at 1.50) earlier, and server 1 will complete customer 1 later (at 3.00), server 2 will get 3.138742 rest while server 1 gets 0 rest.

This means that if you are looking to store the rest property in each customer, you will fail the test case. Because even though the nth customer has a smaller id or comes first chronologically doesn't mean they will get the nth rest time. The rest time is dished out chronologically based on which server completes their customer and is going to enjoy their break now.

**Comment by teominghuang** [comment]
@PokezardVGC Take note, I just re-editted my comment.

**Comment by CalebLee** [comment]
Yup Ming Huang is right, the point of Level 3 is that the rest times are not associated with Customer, but stored somewhere (personally I have like a "Shop" class to manage all the Servers to allocate rest durations to them). The rest times are then assigned to the Servers when they have their break, so you should only .get() the rest duration when the Rest Event occurs.

The same also goes for Service duration for example. Initially they are assigned to each Customer but in Level 5 they are assigned to the Customer only when they are being Served (ie. they could have waited for some time before being Served, so  again the n-th Customer to arrive may not be the n-th Customer to be Served). 

This will be helpful in later levels as well (as I realised), so make sure you allow for this :')

**Comment by PokezardVGC** [comment]
Thanks, peeps, made some minor changes. Btw, just another possible soln for this lv, with minimum effort.
If the currently polled out event is of server-class (we call it Event e)
 - poll out the whole pq into a list
 - iterate through list (iteration at event i)
 - if e's server is the same event i's
 - increment event i's timing
 - throw everything back into pq once done

**Comment by wjingen** [comment]
> Thanks, peeps, made some minor changes. Btw, just another possible soln for this lv, with minimum effort. If the currently polled out event is of server-class (we call it Event e)
> 
> * poll out the whole pq into a list
> * iterate through list (iteration at event i)
> * if e's server is the same event i's server
> * increment event i's timing
> * throw everything back into pq once done

Hi @PokezardVGC,

I tried doing this method but it threw a concurrent modification exception. Probably because I'm removing the event and adding back a new event with the changed timing. 

Any workaround to this? Thanks.

**Comment by wjingen** [comment]
> Thanks, peeps, made some minor changes. Btw, just another possible soln for this lv, with minimum effort. If the currently polled out event is of server-class (we call it Event e)
> 
> * poll out the whole pq into a list
> * iterate through list (iteration at event i)
> * if e's server is the same event i's server
> * increment event i's timing
> * throw everything back into pq once done

Hi @PokezardVGC 

According to #242, I shouldn't be creating a Serve event after encountering a Wait event. Does this suggestion violate what #242 is saying? Thanks.

**Comment by PokezardVGC** [comment]
Not really, unless ur Done event creates a Serve event? Just gotta make sure no cyclic dependency

**Comment by wjingen** [comment]
> Not really, unless ur Done event creates a Serve event? Just gotta make sure no cyclic dependency

HI @PokezardVGC ,

From what I gather from your suggestion, I should create Serve events after encountering Wait events, and when I encounter a ServerRest event I should recreate the priority queue, changing all Serve events' timings. Is that right?

Thanks.

**Comment by PokezardVGC** [comment]
No really, use back the same pq. Basically, empty out the pq, check if any serve events have the same server, if have increased their wait time. Once all these events are done then add them all back into the original pq.

**Comment by wjingen** [comment]
> No really, use back the same pq. Basically, empty out the pq, check if any serve events have the same server, if have increased their wait time. Once all these events are done then add them all back into the original pq. No rest event is created in the process

Hi @PokezardVGC 

In that case it does seem like you're creating Serve events directly after Wait events, contrary to #242 . Is that so?
